apigee/analytics/analytics.go

```diff
@@ -16,96 +16,386 @@ package analytics
+}
+
+// push sends records to UAP.
+func (m *Manager) push(rd *recordData, filename string) error {
+	url, err := m.signedURL(rd, filename)
+	if err != nil {
+		return fmt.Errorf("signedURL: %s", err)
+	}
+
+	b, err := base64.StdEncoding.DecodeString(rd.EncodedRecords)
+	if err != nil {
+		return fmt.Errorf("base64 decode: %s", err)
+	}
+
+	buf := bytes.NewBuffer(b)
+	req, err := http.NewRequest("PUT", url, buf)
+	if err != nil {
+		return fmt.Errorf("http.NewRequest: %s", err)
+	}
+
+	req.Header.Set("Expect", "100-continue")
+	req.Header.Set("Content-Type", "application/x-gzip")
+	req.Header.Set("x-amz-server-side-encryption", "AES256")
+	req.ContentLength = int64(buf.Len())
+
+	resp, err := m.client.Do(req)
+	if err != nil {
+		return fmt.Errorf("client.Do(): %s", err)
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != 200 {
+		return fmt.Errorf("push to store returned %v", resp.Status)
+	}
+	return nil
+}
+
+// signedURL constructs a signed URL that can be used to upload records.
+func (m *Manager) signedURL(rd *recordData, filename string) (string, error) {
+	req, err := http.NewRequest("GET", m.analyticsURL, nil)
+	if err != nil {
+		return "", err
 	}
 
-	body := new(bytes.Buffer)
-	json.NewEncoder(body).Encode(request)
+	q := req.URL.Query()
+	q.Add("tenant", fmt.Sprintf("%s~%s", rd.Org, rd.Env))
+	q.Add("relative_file_path", path.Join(m.uploadDir(), filename))
+	q.Add("file_content_type", "application/x-gzip")
+	q.Add("encrypt", "true")
+	req.URL.RawQuery = q.Encode()
+
+	req.SetBasicAuth(rd.Key, rd.Secret)
 
-	req, err := http.NewRequest(http.MethodPost, axURL.String(), body)
+	resp, err := m.client.Do(req)
 	if err != nil {
+		return "", err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return "", fmt.Errorf("non-200 status returned from %s: %s", m.analyticsURL, resp.Status)
+	}
+
+	var data struct {
+		URL string `json:"url"`
+	}
+	if err := json.NewDecoder(resp.Body).Decode(&data); err != nil {
+		return "", fmt.Errorf("error decoding response: %s", err)
+	}
+	return data.URL, nil
+}
+
+// uploadDir gets a directory for where we should upload the file.
+func (m *Manager) uploadDir() string {
+	now := m.now()
+	d := now.Format("2006-01-02")
+	start := now.Unix()
+	end := now.Add(m.collectionInterval).Unix()
+	return fmt.Sprintf(pathFmt, d, start, end)
+}
+
+// SendRecords sends the records asynchronously to the UAP primary server.
+func (m *Manager) SendRecords(ctx *auth.Context, records []Record) error {
+	r, err := buildRequest(ctx, records)
+	if r == nil || err != nil {
 		return err
 	}
 
-	req.SetBasicAuth(auth.Key(), auth.Secret())
-	req.Header.Set("Content-Type", "application/json")
-	req.Header.Set("Accept", "application/json")
+	for _, record := range records {
+		if err := m.validate(record); err != nil {
+			return fmt.Errorf("validate(%v): %s", record, err)
+		}
+	}
+
+	// Encode records into gzipped JSON
+	buf := new(bytes.Buffer)
+	b64 := base64.NewEncoder(base64.StdEncoding, buf)
```

Comment:
Is there a way we could stream this directly to the file instead of in-memory?

Reply:
Nope, it is not a gzip file that gets written, instead it is JSON that has one field of gzipped data. This is because we need a few fields like key/secret in order to do the upload, but we don't have those fields available in the goroutine that does the upload.

An alternative is to not do any gzipping here, and do the gzipping when uploading.

Reply:
Oh... Speaking of which, we're creating and saving a new file each time this is called. I'm investigating, but I have feeling the calls are going to be too rapid for that and we'll need to have more of an append to log and rollover strategy.